lume_model/variables.py

```diff
@@ -1,102 +1,128 @@
+Variables are designed as pure descriptors and thus aren't intended to hold actual values,
+but they can be used to validate encountered values.
 
 For now, only scalar variables (floats) are supported.
 """
-import logging
-from typing import Optional, Generic, TypeVar
-from pydantic import BaseModel, Field
+import math
+from abc import ABC, abstractmethod
+from typing import Any, Optional, Type
 
-logger = logging.getLogger(__name__)
+from pydantic import BaseModel, field_validator, model_validator
 
 
-# define generic value type
-Value = TypeVar("Value")
-
-
-class Variable(BaseModel, Generic[Value]):
-    """
-    Minimum requirements for a variable.
+class Variable(BaseModel, ABC):
+    """Abstract variable base class.
 
     Attributes:
         name: Name of the variable.
-        value: Value assigned to the variable.
-        precision: Precision to use for the value.
     """
     name: str
-    value: Optional[Value] = None
-    precision: Optional[int] = None
-
-
-class ScalarVariable(BaseModel):
-    """
-    Base class used for constructing a scalar variable.
-
-    Attributes:
-        variable_type: Indicates scalar variable.
-        units: Units associated with scalar value.
-        parent_variable: Variable for which this is an attribute.
-    """
-    variable_type: str = "scalar"
-    units: Optional[str] = None  # required for some output displays
-    parent_variable: str = (
-        None  # indicates that this variable is an attribute of another
-    )
-
-
-class InputVariable(Variable, Generic[Value]):
-    """
-    Base class for input variables.
-
-    Attributes:
-        default: Default value assigned to the variable.
-        is_constant: Indicates whether the variable is constant.
-    """
-    default: Value  # required default
-    is_constant: bool = Field(False)
 
+    @property
+    @abstractmethod
+    def default_validation_config(self) -> dict[str, bool]:
+        """Determines default behavior during validation."""
+        return {}
 
-class OutputVariable(Variable, Generic[Value]):
-    """
-    Base class for output variables. Value and range assignment are optional.
+    @abstractmethod
+    def validate_value(self, value: Any, config: dict[str, bool] = None):
+        pass
 
-    Attributes:
-        default: Default value assigned to the variable.
-        value_range: Acceptable range for value.
-    """
-    default: Optional[Value] = None
-    value_range: Optional[list] = Field(None, alias="range")
+    def model_dump(self, **kwargs) -> dict[str, Any]:
+        config = super().model_dump(**kwargs)
+        return {"variable_class": self.__class__.__name__} | config
 
 
-class ScalarInputVariable(InputVariable[float], ScalarVariable):
-    """
-    Variable used for representing a scalar input. Scalar variables hold float values.
-    Initialization requires name, default, and value_range.
+class ScalarVariable(Variable):
```

Comment:
Would it be clearer to just name it "FloatVariable" since we are hardcoding the scalar type as float across the modules?